### PR TITLE
Add MaskRCNN SwinTransformer-T

### PR DESCRIPTION
**Pull request by eugene123tw**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by sungmanc — DISMISSED** [review]
Generally looks good to me. I have two questions,

1. It seems that `inference_batch_size > 1` can't be applied to all models, is it right? (i.e. resnet50_maskrcnn still used inference batch size as 1). If so, how about adding assert for models that couldn't use the batch inference?
2. Do we still need the `experimental` suffix for the templates? It seems that https://github.com/openvinotoolkit/training_extensions/pull/2362 resolved _template suffix.

File: src/otx/algorithms/detection/adapters/mmdet/task.py
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/task.py:1]: Could you move patching code into `patch_from_hyperparams`? It's invoked at `MMDetectionTask._init_task()`.

File: src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py [hunk 1]
```diff
@@ -316,3 +316,14 @@ def patch_ir_scale_factor(deploy_cfg: ConfigDict, hyper_parameters: DetectionCon
                 ConfigDict(opt_shapes=ConfigDict(input=[1, 3, ir_input_shape[2], ir_input_shape[3]]))
             ]
             print(f"-----------------> x {tile_ir_scale_factor} = {ir_input_shape}")
+
+
+def patch_samples_per_gpu(loader_cfg: ConfigDict, hyper_parameters: DetectionConfig):
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py]: Could you add unit test for this?

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/hpo_config.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+metric: mAP
+search_algorithm: asha
+early_stop: None
+hp_space:
+  learning_parameters.learning_rate:
+    param_type: qloguniform
+    range:
+      - 0.0001
+      - 0.01
+      - 0.0001
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/hpo_config.yaml]:
```suggestion
      - 0.00001
      - 0.001
      - 0.00001
```
Could you change values as above?

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/hpo_config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  learning_parameters.batch_size:
+    param_type: qloguniform
+    range:
+      - 2
+      - 6
```
Comment (eunwoosh) [inline src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/hpo_config.yaml]: Is batch size 8 too big to run on general device? If not, how about to change this value to 8?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your effort. I left some comments. Please take a look.

**Comment by eugene123tw** [comment]
> Generally looks good to me. I have two questions,
> 
> 1. It seems that `inference_batch_size > 1` can't be applied to all models, is it right? (i.e. resnet50_maskrcnn still used inference batch size as 1). If so, how about adding assert for models that couldn't use the batch inference?
> 2. Do we still need the `experimental` suffix for the templates? It seems that [Add model category attributes to model template #2362](https://github.com/openvinotoolkit/training_extensions/pull/2362) resolved _template suffix.

@sungmanc Thanks for reviewing. And please check my response.

1. I prefer to avoid introducing extra checks and assertions unless they are essential for maintaining code integrity. Furthermore, the optimal inference batch size relies on the specific models being used and the underlying hardware, putting the responsibility on developers to make informed decisions when configuring this parameter. By refraining from overly restricting these choices, developers are granted the flexibility to fine-tune this parameter according to their specific needs. It is advisable for developers to conduct experiments, determine the most suitable value, and then incorporate it into the model template.

2. I prefer to keep it as it's important to keep consistency with detection models.

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/hpo_config.yaml (hunk 3 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/hpo_config.yaml]: Updated

File: src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py (hunk 1 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py]: Added

File: src/otx/algorithms/detection/adapters/mmdet/task.py
Comment (eugene123tw) [reply src/otx/algorithms/detection/adapters/mmdet/task.py:1]: Moved

**Comment by sungmanc** [comment]
> > Generally looks good to me. I have two questions,
> > 
> > 1. It seems that `inference_batch_size > 1` can't be applied to all models, is it right? (i.e. resnet50_maskrcnn still used inference batch size as 1). If so, how about adding assert for models that couldn't use the batch inference?
> > 2. Do we still need the `experimental` suffix for the templates? It seems that [Add model category attributes to model template #2362](https://github.com/openvinotoolkit/training_extensions/pull/2362) resolved _template suffix.
> 
> @sungmanc Thanks for reviewing. And please check my response.
> 
> 1. I prefer to avoid introducing extra checks and assertions unless they are essential for maintaining code integrity. Furthermore, the optimal inference batch size relies on the specific models being used and the underlying hardware, putting the responsibility on developers to make informed decisions when configuring this parameter. By refraining from overly restricting these choices, developers are granted the flexibility to fine-tune this parameter according to their specific needs. It is advisable for developers to conduct experiments, determine the most suitable value, and then incorporate it into the model template.
> 2. I prefer to keep it as it's important to keep consistency with detection models.

Thanks for replying.

**Review by sungmanc — DISMISSED** [review]
LGTM

File: src/otx/algorithms/detection/adapters/mmdet/task.py
Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/task.py:1]: Sorry if my word is not clear. I just want to say moving `patch_samples_per_gpu` into `patch_from_hyperparams`. Because both of them patches value according to learning paramaters and `patch_from_hyperparams` is a superset of `patch_samples_per_gpu`.

File: src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py [hunk 4]
```diff
@@ -316,3 +316,20 @@ def patch_ir_scale_factor(deploy_cfg: ConfigDict, hyper_parameters: DetectionCon
...
+        hyperparams (DetectionConfig): OTX detection hyper parameters
+    """
+    params = hyperparams.learning_parameters
+    hparams = ConfigDict(
+        data=ConfigDict(
+            val_dataloader=ConfigDict(samples_per_gpu=int(params.inference_batch_size)),
+            test_dataloader=ConfigDict(samples_per_gpu=int(params.inference_batch_size)),
+        ),
+    )
+    cfg.merge_from_dict(hparams)
```
Comment (jaegukhyun) [inline src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py]: I think this should be moved to patch_from_hyperparms(https://github.com/openvinotoolkit/training_extensions/blob/b9645dc44b6ce66db33072f3939e1920b2a8c6ff/src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py#L618)

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/data_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+
+# pylint: disable=invalid-name
+
+__img_norm_cfg = dict(mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+
+__img_size = (1344, 1344)
```
Comment (jaegukhyun) [inline src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/data_pipeline.py:21]: Is it from mmdet's official codes? It seems too large.

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            nms_pre=1000,
+            max_per_img=1000,
+            nms=dict(type="nms", iou_threshold=0.8),
+            min_bbox_size=0,
+        ),
+        rcnn=dict(score_thr=0.05, nms=dict(type="nms", iou_threshold=0.7), max_per_img=500, mask_thr_binary=0.5),
+    ),
+)
+
+evaluation = dict(interval=1, metric="mAP", save_best="mAP", iou_thr=[0.5])
```
Comment (jaegukhyun) [inline src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py:131]: Why should we use different evaluation config for this model?

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    paramwise_cfg=dict(
+        custom_keys=dict(
+            absolute_pos_embed=dict(decay_mult=0.0),
+            relative_position_bias_table=dict(decay_mult=0.0),
+            norm=dict(decay_mult=0.0),
+        )
+    ),
+)
+
+lr_config = dict(min_lr=1e-08)
```
Comment (jaegukhyun) [inline src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py:147]: Does this lr_config affects to training? We will override lr from template.

File: src/otx/algorithms/detection/utils/utils.py [hunk 8]
```diff
@@ -289,20 +303,14 @@ def create_mask_shapes(
...
+                        points = [Point(x=(point[0][0]) / width, y=(point[0][1]) / height) for point in contour]
 
                     polygon = Polygon(points=points)
                     if cv2.contourArea(contour) > 0 and polygon.get_area() > 1e-12:
                         shapes.append(Annotation(polygon, labels=assigned_label, id=ID(f"{label_idx:08}")))
             else:
-                ellipse = Ellipse(
-                    (box[0] + left) / width, (box[1] + top) / height, (box[2] + left) / width, (box[3] + top) / height
-                )
+                ellipse = Ellipse((box[0]) / width, (box[1]) / height, (box[2]) / width, (box[3]) / height)
```
Comment (jaegukhyun) [inline src/otx/algorithms/detection/utils/utils.py:314]: Are changes for this file for limiting mask size under image size?

**Review by jaegukhyun — COMMENTED** [review]
Generally LGTM, I left some comments.

File: src/otx/algorithms/detection/adapters/mmdet/task.py
Comment (eugene123tw) [reply src/otx/algorithms/detection/adapters/mmdet/task.py:1]: It will not work as other tasks also calls `patch_from_hyperparams` but does not support samplers_per_gpu. For instance, action recognition doesn't use `samplers_per_gpu`, it uses `sampler_per_video`. It fails during the data loader building stage as it doesn't recognize these args.

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/data_pipeline.py (hunk 5 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/data_pipeline.py:21]: It's larger and not the default size, that said, the experiment result is better than MaskRCNN-R50, and this PR is intended to revamp our most accurate model. You can refer to the counting Excel sheet for the numbers of speed/accuracy/GPU usage.

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py (hunk 7 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py:147]: Does the template overwrite the minimum allowed LR in Early Stopping?

https://github.com/openvinotoolkit/training_extensions/blob/b9645dc44b6ce66db33072f3939e1920b2a8c6ff/src/otx/algorithms/common/adapters/mmcv/utils/config_utils.py#L602

Comment (jaegukhyun) [reply src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py:147]: Ah, I thought it is learning rate. Sorry for confusion.

File: src/otx/algorithms/detection/adapters/mmdet/task.py
Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/task.py:1]: I see. I think we need to enable other tasks to support changing inference batch size like a detection later. We can move this function at that step.

File: src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py (hunk 6 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/configs/instance_segmentation/maskrcnn_swin_t/model.py:131]: Every inst-seg model computes IoU:50% as IoU at 0.5:0.95:0.05 took too long to compute. I haven't checked if this is still the case after we optimize the mask evaluation. Maybe we can do some comparison and reinstate it back for future improvement.

https://github.com/openvinotoolkit/training_extensions/blob/2e41196fa37dd3d38edbc50393e948cbccb3eb3d/src/otx/algorithms/detection/configs/instance_segmentation/resnet50_maskrcnn/model.py#L162

File: src/otx/algorithms/detection/utils/utils.py (hunk 8 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/utils/utils.py:314]: It's not limiting mask sizes but preventing potential mask-shifting errors. Previously we crop masks according to boxes and then shift the coordinates, however, I found discrepancies and it's error-prone, so I changed this to align the mask post-processing with OV Model API for mask post-processing (also mitigating potential accuracy differences PyTorch vs. OV): 

https://github.com/openvinotoolkit/model_api/blob/b054796d1900e9b76fe113f756462077294f953d/model_api/python/openvino/model_api/models/instance_segmentation.py#L259

File: src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py (hunk 4 above)
Comment (eugene123tw) [reply src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py]: I tried and it doesn't work as other tasks also calls `patch_from_hyperparams` but does not support `data.val_dataloader.samples_per_gpu` or `data.test_dataloader.samples_per_gpu`. For instance, action recognition uses `samples_per_video`. It fails during the data loader building stage as it doesn't recognize these args.

Comment (jaegukhyun) [reply src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py]: How about adding branch such as if getattr(params, "inference_batch_size", None) before adding val_dataloader and test_dataloader to config

Comment (eugene123tw) [reply src/otx/algorithms/detection/adapters/mmdet/utils/config_utils.py]: I'm uncertain about incorporating this branch, as it implies that `inference_batch_size` could be universally applied across various tasks. In practice, even if this parameter is present in the action template, its functionality would remain limited due to varying naming conventions. However, I acknowledge your perspective, especially given that only detection and instance segmentation currently offer support for this feature.

Btw, `hasattr` doesn't work for action classification:
```python
2023-08-14T09:51:20.3542183Z Traceback (most recent call last):
2023-08-14T09:51:20.3547234Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/bin/otx", line 8, in <module>
2023-08-14T09:51:20.3548530Z     sys.exit(main())
2023-08-14T09:51:20.3550607Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/otx/cli/tools/cli.py", line 77, in main
2023-08-14T09:51:20.3552155Z     results = globals()[f"otx_{name}"]()
2023-08-14T09:51:20.3554332Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/otx/cli/tools/train.py", line 183, in main
2023-08-14T09:51:20.3555823Z     return train(exit_stack)
2023-08-14T09:51:20.3557895Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/otx/cli/tools/train.py", line 276, in train
2023-08-14T09:51:20.3559328Z     task.train(
2023-08-14T09:51:20.3561600Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/otx/algorithms/action/task.py", line 129, in train
2023-08-14T09:51:20.3563220Z     results = self._train_model(dataset)
2023-08-14T09:51:20.3565633Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/otx/algorithms/action/adapters/mmaction/task.py", line 333, in _train_model
2023-08-14T09:51:20.3567267Z     train_model(
2023-08-14T09:51:20.3569334Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/mmaction/apis/train.py", line 220, in train_model
2023-08-14T09:51:20.3571146Z     val_dataloader = build_dataloader(val_dataset, **dataloader_setting)
2023-08-14T09:51:20.3573823Z   File "/home/validation/actions-runner/_work/training_extensions/training_extensions/.tox/tests-act-py310/lib/python3.10/site-packages/mmaction/datasets/builder.py", line 146, in build_dataloader
2023-08-14T09:51:20.3575424Z     data_loader = DataLoader(
2023-08-14T09:51:20.3576746Z TypeError: DataLoader.__init__() got an unexpected keyword argument 'samples_per_gpu'
2023-08-14T09:51:20.4240103Z FAILED
```

To recap, I'll proceed as per your recommendation, but I'll include a note emphasizing that its applicability to other tasks is hindered by divergent naming conventions. If you have any insights on enhancing task compatibility, please do share.

**Review by jaegukhyun — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]
